[IV-25-26]Objetivo4

oscar0310 opened the pull request:
* [ ]  ¿Hay un camino claro que permita ir desde el código hasta la historia de
   usuario que desarrolla, pasando por el mensaje de commit y el issue
   correspondiente?
* [ ]  ¿Se ha intentado cubrir con tests una parte de la lógica de negocio que se
   esté desarrollando?
* [ ]  ¿Hay un producto mínimamente viable que describa lo que se va a entregar en
   este objetivo?
* [ ]  ¿En este producto mínimamente viable se han priorizado unas clases,
   módulos o paquetes más fundamentales frente a otros, movidos a otro
   PMV/Hito?
* [ ] ¿Estoy leyendo todo lo que marco o simplemente marco lo que se me pone por
delante?
* [ ] ¿Se han documentado las elecciones de biblioteca de aserciones y del
   de *test runner*?
* [ ] ¿Te has asegurado que lo que mencionas como biblioteca de aserciones y
test runners lo son realmente y tienen la misma funcionalidad? ¿O estás poniendo
`unittest` como *test runner* y `pytest` como biblioteca de aserciones?
* [ ] ¿Esa documentación incluye criterios de aceptación y también criterios de
búsqueda para las diferentes opciones?
* [ ]  ¿Se han seguido los principios F.I.R.S.T. en la creación del test y
   se ha documentado cómo se ha hecho?
* [ ]  ¿Se describe claramente en el PMV los tests que hay que pasar para que se
   considere viable?
* [ ] ¿Se han testeado también las excepciones?
closes #31 
closes #32 
closes #43
closes #44 
closes #45 
closes #46 
closes #47

oscar0310 committed:
#32 Añadimos el archivo donde se implementran los test

En necesario testear el código que implementamos, porque código no testeadocódigo roto

oscar0310 committed:
#32 Añadiendo la clave test

oscar0310 committed:
#32 añadiendo el archivo para justificar la herramienta para los test

oscar0310 committed:
#32 Añadiendo apartado de Test al README.md

oscar0310 committed:
#32 añadiendo la orden para los test en el readme.md

JJ commented:
@gosema @jorgelopez-ugr por favor revisad este proyecto; aunque no esté listo, mirad a ver si los issues son correctos y demás.

gosema commented:
Buenas tardes @oscar0310 he estado revisando tu código y los issues que has estado asociando a cada commit y tengo varios consejos para ti con el fin de tener que hacer menos correcciones en el futuro.

- Lo primero de todo, te recomendaría que comenzarás a trabar sobre la elección de herramientas de testing ya que es una parte muy importante del objetivo y puedo ir revisándola a la vez que tu avanzas con tu lógica de negocio y creación de tests.
- Después, creo que tu intención de lógica de negocio para tu problema tiene que ver con la obtención de datos en un formato válido y la comprobación de que haya un mínimo de personas en cada sección. Esto lo veo correcto pero creo que los issues asociados no están bien.
- La mayor parte del código lo estas subiendo refiriéndose al issue #16 pero este issue abarca demasiados problemas como rotaciones, dias libres, turnos... Entonces te recomendaría que creases un issue centrado en el problema real y mínimo (quizá el mínimo de personas) y lo completases con como se presentan estos datos y que problemas se presentarán para comprobarlo (#31)
- Si tu intención es que el issue para la lógica de negocio sea #35, creo que habría que hacerse referencia a los commits con este issue y también habría que modificarlo con las pautas que te he dicho. Además, cuando hablas del mínimo de personas en turno de tarde y mañana ya te estarías metiendo en otro problema a parte que por motivos de tiempo de esta asignatura no conviene abordar.
- Por último te recuerdo para evitar estos posibles falllos en el futuro que cuando vayas a crear los tests referidos a la lógica de negocio estén asociados únicamente con el issue #31 o el que uses, ya que los tests deben de salir de forma natural de lo que se describe en el issue. Y no cometer el error de crearlos asociandolos a otro issue como el #32 

Espero que te sirvan estos consejos para corregir lo que llevas y prevenir de errores típicos tu proyecto. Estaré atento cuando tengas una versión más completa y necesites corrección.

oscar0310 commented:
Muchísimas gracias @gosema por el comentario, me has ayudado un montón con este comentario para aclararme un poco.

oscar0310 committed:
#32 Añadiendo justificación de test a nivel de Aserciones

Se justifica que librería a nivel de aserciones vamos a usar para el proyecto

oscar0310 committed:
#32 Cambiando la decisión de test a nivel de decisiones

Voy a seguir mejor como criterio el estandar de facto

oscar0310 committed:
#32 Añadiendo la justificación de elección del framework de tests

oscar0310 committed:
#32 Mejorando la justificación de elección del framework de tests

oscar0310 committed:
#32 Añadiendo la justificación Nivel Herramientas CLI y arreglando errores ortográficos

Con esto completamos la justificación de las herramientas que usaremos para los test y corrijo todos lo errores ortográficos que había

jacarmona364 commented on docs/documentacion_test.md:
```diff
@@ -0,0 +1,43 @@
+Para asegurarnos de que el desarrollo del proyecto cumple todos los deseos del cliente, necesitamos realizar test. Los test son los encargados de verificar que se cumplen estos requisitios. Estas herramientas se dividen en varios niveles de abstracción para ello vamos a ir viendo los diferenetes niveles y nos quedaremos con una herramienta en cada nivel.
+
+## Nivel de Aserciones:
+Este es el nivel más bajo de abstracción, son funciones que se encargan de comparar la salida obtenida con la deseada. Como criterio de elección se busca que la herramienta sea el estándar de facto.
```
Quizás sea conveniente introducir algún criterio más que demuestre que realmente has estudiado las librerías de aserciones y disciernan entre lo que unas y otras tienen y lo que mejor te viene. Se me ocurren criterios en la línea de los principios F.I.R.S.T. , expresividad del código o que la herramienta permita hacer test sencillos (hablando de cantidad de código, no de algo subjetivo). Desde luego lo que mejor iría es tomar ejemplos de aserciones de cada opción y compararlas, de ahí sacarás los mejores criterios.

jacarmona364 commented on docs/documentacion_test.md:
```diff
@@ -0,0 +1,43 @@
+- [RSpec Expectations](https://rspec.info/documentation/3.12/rspec-expectations/) es la librería de aserciones que viene con RSpec y sigue el estilo BDD.
+- [Test::Unit::Assertions](https://test-unit.github.io/test-unit/en/Test/Unit/Assertions.html) es la librería de aserciones que viene con Test::Unit y sigue el estilo clásico de aserciones.
+
+Siguiendo el criterio de elección nos quedamos con [Minitest::Assertions](https://www.ruby-toolbox.com/categories/testing_frameworks).
```
Que el criterio sea tomar el estándar de facto, indica que el proceso se ha realizado a la inversa, se ha escogido la herramienta más estandarizada y se ha justificado con ese mismo criterio. Lo ideal es no empecinarse en una, sino ver ejemplos y analizar la que mejor se adapte a los criterios. Recuerda que no hay elecciones correctas ni incorrectas, mientras la decisión que se tome esté fundamentada, estará 👍

jacarmona364 commented on docs/documentacion_test.md:
```diff
@@ -0,0 +1,43 @@
+Siguiendo el criterio de elección nos quedamos con [Minitest::Assertions](https://www.ruby-toolbox.com/categories/testing_frameworks).
+
+## Nivel Test runners o frameworks:
+Este es el nivel intermedio de abstracción, esta herramienta se encarga de ejecutarlos, agruparlos y planificar los tests. También puede incluir fixtures o funciones para crear objetos que se van a utilizar. Como criterio de eleccion se busca que la herramienta que la herramienta sea el estándar de facto en ruby y que destaque por su velocidad.
```
Cuidado con los typos.

jacarmona364 commented on docs/documentacion_test.md:
```diff
@@ -0,0 +1,43 @@
+
+-[Minitest](https://minite.st/index.html) framework de test que viene por defecto incluida en el lenguaje, destaca por su velocidad.
+-[Rspec-core](https://rspec.info/documentation/3.13/rspec-core/) framework de test muy popular en ruby que viene con el ecosistema RSpec, es bastante más lento que Minitest.
+-[Test::Unit](https://test-unit.github.io/test-unit/en/Test.html) framework clásico de ruby, su rendimiento se situa en un punto intermedio es más lento que Minitest y más rápido que Rspec.
```
Realmente, por mucho que el criterio sea de nuevo que la herramienta sea el estándar y que sea rápida, esta decisión está estrechamente ligada con la biblioteca de aserciones. Estás obligado a coger el test runner para la biblioteca de aserciones seleccionada. Investiga si habría otra forma de solucionar esto o, por el contrario, puede convertirse en un criterio de cara a la elección de la herramienta mientras se fundamente de manera correcta.

jacarmona364 commented on docs/documentacion_test.md:
```diff
@@ -0,0 +1,43 @@
+
+Es el nivel más alto de abstracción, también conocidos como test runners, son los encargados ejecutar los test. Buscan los diferentes scripts que ejecutan los test, analizan la salida y producen un informe.
+
+Como criterio de elección se busca que la herramienta sea el estándar de facto en ruby.
```
De nuevo, vamos a necesitar algo más que ir a lo sencillo, por desgracia. Lo de no necesitar instalaciones adicionales también podría ser un criterio, añádelo si así lo consideras. En cualquier caso, con los comentarios anteriores creo que puedes deducir más o menos lo que necesitas aquí.

jacarmona364 reviewed (commented):
Buenas noches @oscar0310, como veo que estás aún trabajando en el objetivo no voy a evaluar la parte que potencialmente imagino que vas a modificar, trataré de completar un poco más las aportaciones de @gosema para facilitarte el desarrollo.

En primer lugar, acerca de los mensajes de commit, recuerda que para seguir la metodología debes hacer cada commit lo más representativo posible. Te recomendaría que revises las palabras reservadas como fix, chore, feat y demás que sirven para orientar la finalidad del commit. Poner fix: Arreglo ... , related/closes #issue , facilita mucho la revisión posterior de los cambios. 

Por la parte de la elección de herramientas de test, no vas desencaminado pero echa un vistazo a los comentarios acerca de los criterios y los ejemplos prácticos que creo que pueden serte útiles.

Estoy pendiente de tus modificaciones, ánimo.

oscar0310 committed:
fix: Corrigiendo en #32 la justificación de la librería de aserciones

se elimina el criterio de el estandar de facto porque no seguía una correcto desarrollo ya que al establecerlo ya estabas eligiendo directamente  y se añade como criterios que siga TDD y que sea una Herramienta Nativa que evite configuraciones externas

oscar0310 committed:
fix: Corrigiendo en #32 la justificación del nivel del framework

 Se elimina el criterio del estandar de facto ya que no era correcto ya que esta selección está más relaccionada con la librería de aserciones y se usa esto como criterio

oscar0310 committed:
fix: Corrigiendo en #32 la justificación del nivel CLI para ejecutar los test

Se elimina el criterio aquí también del estandar de facto como en los anteriores y se establece de cara al futuro como criterio una herramienta que permita la automatización de los test para siguientes objetivos y queremos evitar la instalación de más gemas externas que provoque configuraciones externas que pueden acarrear deuda técnica.

oscar0310 committed:
fix: Corrigiendo en #32 el formato de la presentación de las herramientas

Se añade un espacio entre ellas fara facilitar la lectura

vpedrosa reviewed (changes requested):
@oscar0310 te dejo una pequeña revisión de los issues, ya que aún te falta el código de la implementación y de los tests y a mí me acaban de hacer una buena revisión con cosas que creo que puedes aplicar. 

Las issues #39 , #40 , #41 y #42 se solapan y podrían ser redundates.

Veo que te falta el "camino infeliz" en los issues, dicen qué debe cumplirse pero no definen qué pasa cuando no se cumple.

oscar0310 committed:
fix: #44 definiendo la función comprobar_ext

Se define esta función para comprobar que el archivo sea en formato CSV, sin esta comprobación se podría intentar usar otras extensiones las cuales pueden provocar que no funcione correctamente la librería de creación de horarios

oscar0310 committed:
fix: #45 definiendo la función comprobar_nTrabajador

Se define esta función para comprobar que el nombre de un trabajador sigue el formato correcto, siendo así un trabajador válido

oscar0310 committed:
fix: #46 Definiendo la función comprobar_Seccion

Definimos esta función porque es necesario comprobar que la sección de cada trabajador sea válida y forma parte de las secciones definidas por el modelo del problema ya que si no forma parte ese trabajador no será válido.

oscar0310 committed:
fix: #43 Creando método leer_archivo

Se crea este método porque sin el no podemos leer la información del archivo, se encarga solo de leer todas las lineas menos la primera ya que contiene en ella Nombre y Sección como se describe en el ejemplo del user-journey, además cumple el principio SOLID ya que solo tienen una funcionalidad

oscar0310 committed:
fix: #43 Creando la función procesar_datos

Se crea el método encargado de obtener los datos de cada trabajador del archivo, se lee cada linea del archivo para obtener la información, la cual esta función se delega en otra función para no romper el principio de única funcionalidad y este se añade en un set para que no haya datos repetidos

oscar0310 committed:
fix: #43 Comenzamos a definir el método obtener_datos_trabajador

Éste método es el encargado de devolver los datos de cada trabajador al método procesar_datos, recibe la linea la cual se parte por la coma al ser un csv y se obtiene nombre y sección, la sección se tranforma a símbolo porque si no no se puede usar para compararla con el enum de Secciones creado en la modelización del problema. Y se comprueba los datos para que sean validos

oscar0310 committed:
fix: #31 Hacemos únicos a los trabajadores

De cara a hacerlos únicos para que se puedan comparar generámos un identificador para cada uno, este irá ligado al objeto trabajador para hacerlos comparables, para ello en procesar datos lo creamos con una solución muy ruby, que es darle a cada línea un indice file.each_with_index que crea un índice que se usará encomo id. Este id se pasa al obtener_datos_trabajador ya que es donde se van a procesar todos los datos de los trabajadores y se asocirá allí ese id a un trabajador, y este será devuelto junto a los datos del trabajador

oscar0310 committed:
fix: #47 Creamos un constructor para la clase  trabajador

Se crea el constructor para poder inicializar los objetos de los trabajadores, permitiendo así poder usar la información guardada

oscar0310 committed:
fix: #43 Terminamos la función obtener_datos_trabajador

Tras poder inicializar los objetos trabajadores y disponer de un identificador único inicializamos el objeto trbajador propio porque sin él no podríamos crear un horario, y devolvemos el id, trabjador y la seccion que es lo que añadirémos al set de trabajadores

oscar0310 committed:
fix: #43 Añadimos el archivo trabajador.rb

 Se añade de cara a poder trabajar con el constructor de trabajador

oscar0310 committed:
fix: #43 Añadimos las funciones propias que comprueban el número mínimo de personas

Se crea la función comprobar_num_trabajadores que es la encargada de recorrer cada sección y pasarsela a la función propiamente que comprueba que en esa sección haya un número mínimo de personas para ello establezco que se necesitaría 5 personas ya que se necesitan 2 para el turno de mañana y otras dos para el turno de tarde y mínimo se libra un día más el domingo que cierra la tienda. Se define tambien un error propio de este requisito para que se entienda mejor a la hora de depurar errores.

oscar0310 committed:
fix: #43 Añadiendo la función obtener_datos

Esta función es la encargada de llamar a todas las funciones que hemos implementado para depurar la entrada y garantizar que se devuelva los trabajadores mínimos para cumplir todos los requisitos del horario

oscar0310 committed:
fix: #32 corrigiendo la justificación del framework Test::Unit

Se explica mejor usando el criterio establecido anteriormente

oscar0310 committed:
fix: #43 Arreglando errores de sintaxis

 cambio SECCIONES_VALIDAS por secciones que es cómo esta declarado y quito trbjdor por trabajador

oscar0310 committed:
fix: #32 Degino la tarea de los test en el rakefile

Se define la tarea que se va a usar para poder ejevutar los test

oscar0310 committed:
fix: #43 Arreglando fallos

Se cambia require dominio/trabajador por require_relative dominio_trabajador ya que con require lo busca en las librerías de gemas de ruby y nunca va a encontrarlo. Además se elimina la variable file ya que no se usaba para nada se devuelve directamente el resultado

oscar0310 committed:
fix: #32 añadimos minitest al Gemfile

Para poder ejecutar los test usando bundler para que se tenga en cuenta las dependencias necesitamos que este en el Gemfile. Sin esto bundler no lo tendrá en cuenta y dará error

oscar0310 committed:
fix: #32 Añadimos la gema de Minitest al archivo de los Test

Sin ella no se podrán lanzar los test

oscar0310 committed:
fix: #32 Añadimos un reporter personalizado para minitest

Las salidas de minitest son muy básicas y no son muy autoexplicativas, es por esto que creo un ReporterPersonalizado que es el encargado de modelar la salida en el se crean dos métodos. start que es el encargado de mostrar la primera linea al iniciar los test y record que se encarga de modelar el nombre del test ya que usaré nombres autoexplicativos y los deja como una sentencia limpia. Por ejemplo si el test se llama test_validacion_de_archivo_entrada_csv, se dejará como nombre en la salida de los test: validacion archivo de entrada csv. Después comprueba si lo ha pasado o no y en cada caso se añade un emoji correspondiente para que sea más visual. En caso de no haberlo pasado se muestra el error junto al test incorrecto

oscar0310 committed:
fix: #32 Creando un plugin personalizado con mi reporter

Para que minitest use mi reporter y no use los que tiene propios ya por defectro que son muy básicos, se define el plugin dentro del módulo de Minites para que tenga acceso y lo que se hace es limpiar la lista de los reporters que tiene y se crea el ReporterPersonalizado. Fuera del módulo se hábilita que se use este reporter

oscar0310 committed:
fix: #32 Añadiendo la clase donde estarán todos los test

Se añade esta clase que será la encargada de tener todos los test por esto hereda de Minitest::Test

oscar0310 committed:
fix: #44 Añadiendo test_acepta_archivo_csv

Se realiza este test para comprobar el Happy path en el caso de que el archivo sea un csv, si no lo es dará error

oscar0310 committed:
fix: #44 Añadiendo test_rechaza_archivo_no_csv

Se añade este test para comprobar el Sad path cuando el archivo no es csv, para ello se comprueba que devuelva el error correspondiente

oscar0310 committed:
fix: #45 Añadimos test_nombre_trabajador_valido

 Este test comprueba el Happy path de la función para comprobar que el nombre de los trabajadores sea válido se usa varios nombres que deben aceptar con formatos distintos

oscar0310 committed:
fix: #45 Arreglando errores de comprobar_nTrabajador

No se había tenído en cuenta que la persona puede tener dos nombres como María José y tampóco se había tenido en cuenta que podría comenzar el nombre o los apellidos con una mayuscula con tilde como mi propio nombre Óscar

oscar0310 committed:
fix: #45 definiendo el test_nombre_trabajador_no_valido

En caso de que el trabajador no sea válido la función que lo comprueba debe devolver un error, aquí se comprueba que se devuelva ese error. Si no se devuelve pos fallará el test

oscar0310 committed:
fix: #46 Definimos el test_seccion_valida

Este test se define para comprobar que correctamente acepta a las secciones que se definieron en el objetivo 2, para ello se pasan en formato símbolo ya que se definieron así en el enum SECCIONES

oscar0310 committed:
fix: #45 definiendo el test_seccion_no_valida

Este test será el encargado de comprobar cuando la entrada de una sección no forme parte de las secciones definidas en el enum SECCIONES. Es válido cuando devuelve el error correspondiente a este apartado

oscar0310 committed:
fix: #45 Soluccionando error en comprobar_Seccion

Se cambia en el error el nombre del enum a SECCIONES que es como se llama

oscar0310 committed:
fix: #43 definiendo test_leer_archivo devuelve_un_numero_correcto_lineas

Se crea un test para comprobar que la funcion leer_archivo funciona correctamente, para ello usamos los tres archivos uno con 25 trabajadores, otro con 12 y otro vacío, para ello lo que tiene que devolver es un array con las lineas de ese trabajador y tiene que que eliminar la primera línea porque no es necesaria, es por eso que se comprueba con el tamaño esperado en el test. Los archivos que se suben son las entradas mencionadas previamente con distintos número de trabajadores. Con ellas comprobamos tanto el Happy path que es que tenga trabajadores como el sad path que es que este vacío

oscar0310 committed:
fix: #43 Asegurandome de que esté vacío el archivo de entrada

oscar0310 committed:
fix: #43 Deshaciendo test que se ha subido por error

En el commit anterior he subido por error un test de una función sin querer y no esta justificada la borro para subirla con su justificación correspondiente

oscar0310 committed:
#43 defino el test_obtener_datos_trabajador_devuelve_objeto_correcto

Se crea este test para comprobar el funcionamiento del happy path de la función obtener_datos_trabajador, se espera que devuelva un objeto correspondiente a la entrada si no dará error

oscar0310 committed:
fix: #43 definiendo el test_obtener_datros_trabajador_entrada_invalida

Se define este test para comprobar que efectivamente en obtener_datos_trabajador cuando se añade una entrada inválida devuelve error en lugar de los datos de un trabajador que no es válido

oscar0310 committed:
fix: #43 definiendo los test de la función procesar_datos

Se definen tres test, el primero test_procesar_datos_devuelve_conjunto_correcto se usa para comprobar el Happy path para ello se le  pasa un archivo leido y tiene que almacenar cada línea de información en un set de trabajadores es por eso quese comprueba que la salida se han alamacenado correctamente cada línea en el set. EL segundo test_procesar_datos_devuelve_conjunto_vacio, nos encargamos de comprobar cuando el archivo leido no tiene ni una linea de información, se devuelva un set vacío y no tenga ninguna información basura. EL último test se encarga de comprobar el sad path, se comprueba que efectivamente falle cuando el nombre del trabajador o la sección no es válida.\n Además al haber ejecutado los test me doy cuenta que en la función procesar_datos tengo en la inicialización del set unos parentesis innecesarios que elimino.

oscar0310 committed:
fix: #43 Definiendo los test de la función comprobar_num_TrabajSeccion

Se definen tres test para comprobar el correcto funcionamiento de la función. Para ello comprobamos el happy path con test_comprobar_numero_trabajadores_en_seccion_valido que comprueba que efectiamente la sección pasada contenga 5 personas como definímos cuando creamos la función para satisfacer los requisitos de mínimo de personas por sección. Con test_comprobar_numero_trabajadores_en_seccion_con_entrada_vacia se comprueba que efectivamente falle con un error de que no hay el número minimo de personas ya que al no disponer de trabajadores no podemos satisfacer este requisito y por último con test_comprobar_numero_trabajadores_en_seccion_con_menos_de_5 compobamos que si la sección no tiene ese número de personas da error como el anterior

oscar0310 committed:
fix: #24 definiendo test para la función comprobar_num_Trabajadores

Se crean estos test para comprobar el correcto funcionamiento de la función. Para ello se comprueba el Happy path con test_comprobar_numero_trabajadores_totales_validos, para ello creamos un set que tenga 5 trabajadores por cada sección como definímos cuando implementamos las funciones para cubrir todos los requisitos y comprobamos que no de ningún error. Para comprobar el sad path usamos test_comprobar_numero_trabajadores_totales_con_entrada_vacia donde se busca testear que cuando no hay ningún trabajador de error, ya que no podríamos crear un horario y con test_comprobar_numero_trabajadores_totales_invalido  que comprueba que falla cuando en una sección no disponemos de un número total de 5, ya que sin esos trabajadores no podríamos cumplir los requisitos para hacer un horario correcto

oscar0310 committed:
fix: #43 definiendo test para la función obtener datos

Se realizan estos test para comprobar el correcto  funcionamiento de obtener_datos que es la función encargada de obtener todos los datos depurados, devolviendo el set de trabajadores valido para hacer el horario. Para ello hemos testeado tanto el happy path como el sad path. Para el happy path usamos test_obtener_datos_validos aquí pasamos un archivo que cumple todos los requisitos de entrada para comprobar que devuelve correctamente en la salida el set con todos los trabajadores. En cuanto al sad path usamos test_obtener_datos_con_entradas_invalidas se comprueba que se trate correctamente con todos los fallos posibles que puede ocurrir y que los trata correctamente devolviendo su error correspondiente es por esto que se prueba con archivos que contenga un nombre no valido, una sección no valida, que no tengan un número mínimo de trabajadores por sección, que esté vacío y con un archivo que no es .csv. Para la comprobación de entrada con nombre erroneo y seccion erronea se añade dos .csv con esas características.

oscar0310 committed:
fix: #43 Añadiendo archivos de entrada para testear obtener_datos

Se añade los dos archivos para testear en el caso de que el nombre o la sección no sea válida

oscar0310 commented:
@gosema cuando puedas me das feedback sobre mi objetivo porfa.

gosema commented on docs/documentacion_test.md:
```diff
@@ -0,0 +1,50 @@
+
+En ruby tenemos varias opciones en cuanto a librería de aserciones:
+
+- [Minitest::Assertions](https://docs.ruby-lang.org/en/2.1.0/MiniTest/Assertions.html) es la librería de aserciones que viene por defecto con Minitest esta sigue el estilo clásico de las aserciones como podemos ver en funciones como assert_equal(exp, act, msg = nil) o assert_empty(obj, msg = nil). Esta librería no requiere instalar gemas externas adicionales ya que viene incluida en la librería e ...
```
¿No crees que el enlace de la documentación es un poco anticuado? Es de la 2.1, creo que Ruby ya va por la 4. algo. ¿No hay una fuente más actualizada?

gosema commented on docs/documentacion_test.md:
```diff
@@ -0,0 +1,50 @@
+## Nivel Test runners o frameworks:
+Este es el nivel intermedio de abstracción. Esta herramienta se encarga de ejecutarlos, agruparlos y planificar los tests. También puede incluir fixtures o funciones para crear objetos que se van a utilizar. 
+
+Como criterio de eleccion, buscamos  que la herramienta se integre correctamente con la librería de aserciones sin necesidad de configuraciones complejas y que destaque por su velocidad, mejorando la experiencia del programador cuando la use.
```
No creo que la velocidad sea un buen criterio objetivo ¿tienes algún benchmark que lo pruebe? O quizá hay otro criterio que valga más la pena que este

oscar0310 replied:
En cuanto a la velocidad lo que busco aquí es una herramienta rápida a la hora de realizar los test, evitando perder tiempo cada vez que se lancen. No he usado un benchmark pero uso la documentación propia de minitest sobre comparaciones que han realizado [ellos](https://minite.st/comparisons.html), ¿ Con eso no sería válido para justificarlo?

gosema commented on docs/documentacion_test.md:
```diff
@@ -0,0 +1,50 @@
+## Nivel Test runners o frameworks:
+Este es el nivel intermedio de abstracción. Esta herramienta se encarga de ejecutarlos, agruparlos y planificar los tests. También puede incluir fixtures o funciones para crear objetos que se van a utilizar. 
+
+Como criterio de eleccion, buscamos  que la herramienta se integre correctamente con la librería de aserciones sin necesidad de configuraciones complejas y que destaque por su velocidad, mejorando la experiencia del programador cuando la use.
```
No se si tener como criterio que sea compatible con tu herramienta de aserciones es correcto. Estás sesgando tu opinión a una única elección que has hecho al principio, quizá conviene que después de ver todas las herramientas posibles y elegir al final la mejor combinación. Imagínate que la mejor herramienta de CLI no es compatible con tus aserciones, ¿lo descartarías por ser incompatible? Trata de estudiar otros criterios como integración o el uso de gemas externas

gosema commented on docs/documentacion_test.md:
```diff
@@ -0,0 +1,50 @@
+
+- [Rake::TestTask](https://ruby.github.io/rake/Rake/TestTask.html) es una herramienta que viene por defecto con el gestor de tareas [Rake](https://ruby.github.io/rake/), el cual ya fue elegido en la documentación del gestor de tareas. Rake::TestTask permite definir tareas para ejecutar los test independientemente del framework de test que se use. Al formar parte de Rake no es necesario instalar n ...
+
+- [m](https://github.com/qrush/m) es una herramienta CLI capaz de ejecutar test en minitest y test::unit. Esta herramienta está diseñada para que una persona ejecute los test manualmente desde la línea de comandos, por tanto no está orientada a la automatización de la ejecución de los test. Además requiere instalar una gema adicional para su uso. 
```
¿Estas seguro que sea una herramienta CLI? Mira el github lleva 5 años sin actualizarse ¿sigue siendo una herramienta en uso?

gosema reviewed (commented):
El código y los issues los veo bastante bien. Y en general la documentación también, corrige los errores que te he dicho y creo que el trabajo va por buen camino

oscar0310 commented:
> El código y los issues los veo bastante bien. Y en general la documentación también, corrige los errores que te he dicho y creo que el trabajo va por buen camino

@gosema muchísimas gracias por el feedback.

oscar0310 committed:
fix: #32 Actualizando la documentación de Minitest::Assertions

Se actualiza la documentación a una más actual, ya que estaba una de la versión 2.1 de ruby

oscar0310 committed:
fix: #32 Eliminando m de las opciones de herramienta CLI

Ya que está en desuso y no parece ser un test runner puro

oscar0310 committed:
fix: #32 Mejorando la Justificación a Nivel de Herramientas CLI

Se añade el caso de ejecución nativa del propio lenguaje ruby ya que aunque no sea una herramienta cli propia, permite la ejecución de test y hay que tenerlo en cuenta

oscar0310 committed:
fix: #32 Cambiando los criterios de la elección del Framework

Establecemos los mismos criterios que en la librería de aserciones para seguir la misma metodología de elección

oscar0310 committed:
fix: #32 corrigiendo typo

oscar0310 commented:
Buenas @gosema y @jacarmona364 , he intentado implementar las correcciones que me habéis hecho si me podéis dar feedback, os lo agradecería un montón.

vpedrosa commented on test/test.rb:
```diff
@@ -0,0 +1,198 @@
+
+  def test_obtener_datos_trabajador_devuelve_objeto_correcto
+      resultado=GeneradorHorario::obtener_datos_trabajador("Juan Pérez, pescadería", 0)
+      assert_equal(0, resultado[:id])
```
Comprobar que resultado[:id] sea 0 cuando se pasó 0 como parámetro no es redundante? No estarías solo que Ruby asigna valores?

vpedrosa reviewed (commented):
Te hago una revisión rápida de lo que he visto.

La clase Horario no tiene tests ni lógica de negocio asociada, y tienes atributos no utilizados de Trabajador (último_día_libre y últimos_tres_turnos) en el dominio que no tienen lógica de negocio asociada, ni justificas que no la tengan.  

En obt_datos.rb:74 haces `if trabajadores_en_seccion.size < 5`, pero en #3 indicas que deben ser dos por turno y en la #35 comentas que son 4 (dos por turno). Te lo pongo por aquí porque no sé si debes cambiar las issues o el código para que sean coherentes.

Has definido una serie de errores en el dominio que tampoco estás utilizando, puede ser un problema del objetivo 2 o de este.

En general, y como me pasó a mi también, tienes mucho código y eso hace difícil la revisión, asegúrate de que el MVP es el mínimo viable si no lo has hecho ya.

oscar0310 commented:
> Te hago una revisión rápida de lo que he visto.
> 
> La clase Horario no tiene tests ni lógica de negocio asociada, y tienes atributos no utilizados de Trabajador (último_día_libre y últimos_tres_turnos) en el dominio que no tienen lógica de negocio asociada, ni justificas que no la tengan.
> 
> En obt_datos.rb:74 haces `if trabajadores_en_seccion.size < 5`, pero en #3 indicas que deben ser dos por turno y en la #35 comentas que son 4 (dos por turno). Te lo pongo por aquí porque no sé si debes cambiar las issues o el código para que sean coherentes.
> 
> Has definido una serie de errores en el dominio que tampoco estás utilizando, puede ser un problema del objetivo 2 o de este.
> 
> En general, y como me pasó a mi también, tienes mucho código y eso hace difícil la revisión, asegúrate de que el MVP es el mínimo viable si no lo has hecho ya.

Buenas @vpedrosa actualmente me estoy centrando como lógica de negocio en obtener un número mínimo de personas válidas para obtener en un futuro el horario, es por eso que la clase horario no tiene implementado nada ya que se usará de cara a crear el horario cómo se justificó en el objetivo 2. Lo mismo pasa con trabajador al tener solo en la entrada el nombre no se tendrá en cuenta nada más para crearlo, último_día_libre y últimos_tres_turnos se crearon en el objetivo 2 de cara a almacenar los ultimos tres turnos del trabajador y el último día liibre, los cuales no las uso ya que se usaran cuando se cree el horario y comience a solucionar los issues #18  #17, ya que son otros problema que no se van a solucionar en este milestone, solo me voy centrar en lo que he hecho.

Lo de los trabajadores se justifica en el commit de cuando creo la función, porque si pusiera el número directamente en el issue estaría dando la solución, este es el [commit](https://github.com/oscar0310/IV25-26/pull/33/changes/3029593b35b711a1cbb2f575bfbae64e3f4be3d9) asociado por si le quieres echar un ojo .

Los errores definidos que no se usan actualmente son también para los issues que no se resuelven en este milestone.

En cuanto a código en si es poco ya que es solo lo de obtener datos y  verificar que son válidos, lo que tiene más lineas  son los test, pero para cumplir lo que se explica en el objetivo 4 sobre que código no testeado es código roto y tengo que comprobar tanto el sad path, como el happy path  no veo manera de reducirlo.

oscar0310 committed:
fix: #24 eliminando comprobación innecesaria de un test

se elimina la comprobación del id de test_obtener_datos_trabajador_devuelve_objeto_correcto ya que lo único que se hace es asignarlo

gosema reviewed (approved):
Veo mucho mejor la documentación y le código me parece correcto. Tienes mi visto bueno

jacarmona364 commented on docs/documentacion_test.md:
```diff
@@ -0,0 +1,50 @@
+Para asegurarnos de que el desarrollo del proyecto cumple todos los deseos del cliente, necesitamos realizar test. Los test son los encargados de verificar que se cumplen estos requisitios. Estas herramientas se dividen en varios niveles de abstracción para ello vamos a ir viendo los diferenetes niveles y nos quedaremos con una herramienta en cada nivel.
+
+## Nivel de Aserciones:
+Este es el nivel más bajo de abstracción, son funciones que se encargan de comparar la salida obtenida con la deseada.
```
¿Consideras necesario para superar el objetivo dar la definición de cada tipo de herramienta o puede ser algo externo al mismo que solo agregue ruido?

jacarmona364 commented on docs/documentacion_test.md:
```diff
@@ -0,0 +1,50 @@
+## Nivel de Aserciones:
+Este es el nivel más bajo de abstracción, son funciones que se encargan de comparar la salida obtenida con la deseada.
+
+Como criterio de elección se busca que la librería de aserciones siga el estilo TDD clásico (assert_equal(esperado,valor_actual)) priorizandolo frente al estilo BDD(expect.(actual).to_eq(experado)) y no se requiera instalar gemas externas adicionales para su uso, evitando tener que añadir archivos de configuración adicionales.
```
Reducir las instalaciones adicionales o las dependencias y configuraciones pertinentes de la herramienta, son un criterio que no va más allá de "ahorrar trabajo". Es evidente que debe ser una de las justificaciones para tomar una decisión, pero no la principal, ya que es por ello por lo que puedes dejarte fuera otras que tengan "cosas" que te estés dejando. Dada la poca variedad de herramientas actualizadas para ruby en comparación con otros lenguajes, te recomiendo que trates de hacer una prueba con cada una y en base a eso elijas un criterio más "tangible" como puede ser lo representativas que sean las sentencias o lo que te permitan minimizar el código para hacerlo más limpio. De esta forma puedes cerciorarte de si realmente, dentro de tus opciones, es esa la mejor o no.

jacarmona364 commented on docs/documentacion_test.md:
```diff
@@ -0,0 +1,50 @@
+
+-[Test::Unit](https://test-unit.github.io/test-unit/en/Test.html) framework clásico de ruby, anteriormente era el estándar en ruby pero actualmente si se quiere usar se necesita instalar gemas externas adicionales ya que ya no viene incluida en la librería estandar de ruby y necesita configuración adicional para su uso. Este framework sigue el estilo clásico de TDD como podemos ver en su document ...
+
+Siguiendo los criterios de elección nos quedamos con Minitest ya que no requiere librerías externas adicionales y sigue el estilo TDD clásico.
```
¿Que sea el test runner de la librería de aserciones elegida tiene también algo que ver con la decisión?

jacarmona364 commented on docs/documentacion_test.md:
```diff
@@ -0,0 +1,50 @@
+
+Como criterio de elección se busca que la herramienta sea capaz de automatizar la ejecución de los test de cara a los proximos objetivos de la asignatura y no se necesite instalar gemas evitando configuraciones innecesarias y reduciendo la deuda técnica.
+
+A la hora de elegir una herramienta CLI para ejecutar los tests en ruby con Minitest tenemos no tenemos una gran variedad de opciones:
```
Ojo con los typos.

jacarmona364 commented on docs/documentacion_test.md:
```diff
@@ -0,0 +1,50 @@
+
+## Nivel Herramientas CLI para ejecutar los tests:
+
+Es el nivel más alto de abstracción, también conocidos como test runners, son los encargados ejecutar los test. Buscan los diferentes scripts que ejecutan los test, analizan la salida y producen un informe.
```
En serio, no son necesarias estas definiciones, y menos si pueden dar lugar a confusiones sobre si comprendes las diferencias entre cada una de las tres herramientas que debes elegir.

jacarmona364 commented on docs/documentacion_test.md:
```diff
@@ -0,0 +1,50 @@
+
+Es el nivel más alto de abstracción, también conocidos como test runners, son los encargados ejecutar los test. Buscan los diferentes scripts que ejecutan los test, analizan la salida y producen un informe.
+
+Como criterio de elección se busca que la herramienta sea capaz de automatizar la ejecución de los test de cara a los proximos objetivos de la asignatura y no se necesite instalar gemas evitando configuraciones innecesarias y reduciendo la deuda técnica.
```
Si el criterio es que no requieran instalaciones externa y pones herramientas que las requieren sí o sí, denota que la elección se ha realizado sabiendo lo que se iba a escoger y justificando esa decisión con otras opciones que no deberían entrar en la discusión.

jacarmona364 reviewed (commented):
Buenas tardes Óscar, te dejo algunos comentarios sobre tu objetivo, en general no lo veo mal encaminado pero pienso que las dificultades vienen de la poca variedad de elección que tienes con ruby, en comparación con otros lenguajes. Esto te obliga a justificar de manera más estricta tu decisión, entrando en comparaciones más detalladas sobre las herramientas. Como veo que estás trabajando bastante y tampoco veo necesario que rehagas todo el trabajo, céntrate en la elección de criterios y la justificación, para que pueda ver que has descartado las demás herramientas a sabiendas de lo que ofrecen o no y tendrás mi aprobado, ánimo.

oscar0310 committed:
fix: #32 Eliminando las descripción de los niveles

 No son necesarias y solo añade ruido

oscar0310 committed:
fix: #32 Corrigiendo la justificación de la herramienta de test

Se sigue las recomendaciones de jcarmona

oscar0310 commented:
@JJ listo para la revisión.

JJ commented on docs/documentacion_test.md:
```diff
@@ -0,0 +1,46 @@
+
+A la hora de elegir una herramienta CLI para ejecutar los test en ruby disponemos de las siguientes opciones:
+
+- Ejecución nativa(Ruby):Aunque no sea una herramienta CLI como tal, ruby permite ejecutar los test directamente desde la línea de comandos usando el intérprete de ruby seguido del script que contiene los test como podemos ver en la documentación de [minitest](https://github.com/minitest/minitest?tab=readme-ov-file#running-your-tests).  No permite automatizar la ejecución de los test, ya que cada ...
```
Pero se puede meter en el gestor de tareas. No hay que hacerlo "manualmente". Para eso están los gestores de tareas.

JJ commented on lib/generador_horario/obt_datos.rb:
```diff
@@ -0,0 +1,78 @@
+        formato3=/^[a-zA-ZáéíóúÁÉÍÓÚñÑ]+\s+[a-zA-ZáéíóúÁÉÍÓÚñÑ]+\s+[a-zA-ZáéíóúÁÉÍÓÚñÑ]+\s+[a-zA-ZáéíóúÁÉÍÓÚñÑ]+$/
+
+        unless nombre=~formato1 || nombre=~formato2 || nombre=~formato3
+            raise ArgumentError, "El nombre: #{nombre} pasado en el archivo de entrada no es un nombre valido, tiene que seguir el formato nombre apellido o nombre apellido1 apellido2"
```
Por favor, comprueba los errores de ortografía como "válido"

JJ commented on lib/generador_horario/obt_datos.rb:
```diff
@@ -0,0 +1,78 @@
+        end
+    end  
+    
+    def comprobar_nTrabajador(nombre)
```
debéis utilizar nombres completos en código limpio, no abreviaturas.

JJ reviewed (approved):
La verdad, si esto lo hubieras presentado en diciembre te habría pedido que le dieras una vuelta, porque para los nombres de trabajadores lo único que tendrías que comprobar es que son únicos, y la lógica de negocio debería ser asignar un mínimo de trabajadores, no comprobar que están asignados. Pero la verdad es que entre los revisores y tú habéis hecho un buen trabajo, y yo creo que no se puede pedir más.

oscar0310 committed:
fix: #32 Actualizando la explicación de Ejecución Nativa(ruby)

Se explica que para automatizar la ejcución de los test dependería de un gestor de tareas

oscar0310 committed:
fix: #45 cambiando el nombre de la función comprobar_nTrabajador

En código limpio no se usan abreviaturas, por tanto se cambia a comprobar_nombre_trabajador

oscar0310 committed:
fix: #45 corrigiendo error ortográfico

oscar0310 committed:
fix: #45 #46 arreglando más errores ortogáficos
